amxmodx/file.cpp

```diff
@@ -7,484 +7,503 @@
 // Additional exceptions apply. For full license details, see LICENSE.txt or visit:
 //     https://alliedmods.net/amxmodx-license
 
-// intptr_t
-#ifdef _MSC_VER
-	typedef int intptr_t;
-	#define _INTPTR_T_DEFINED
-#endif
-
-#ifdef __GNUC__
-#include <stdio.h>
-#include <sys/types.h>
-#include <sys/stat.h>
-#include <dirent.h>
-#endif
-
-// header file for unlink()
-#if defined(__linux__) || defined(__APPLE__)
-#include <unistd.h>
-#else
-#define WINDOWS_LEAN_AND_MEAN
-#include <windows.h>
-#include <io.h>
-#endif
-
 #include "amxmodx.h"
+#include "CFileSystem.h"
+#include "CLibrarySys.h"
 
-CVector<FILE *> FileList;
+using namespace ke;
 
-class AutoFilePtr
+// native read_dir(const dirname[], pos, output[], len, &outlen = 0);
+static cell AMX_NATIVE_CALL read_dir(AMX *amx, cell *params)
 {
-	FILE *m_FP;
-public:
-	AutoFilePtr(FILE *fp) : m_FP(fp)
-	{}
-	
-	~AutoFilePtr()
-	{
-		if (m_FP)
-			fclose(m_FP);
-	}
-	
-	operator FILE* ()
-	{
-		return m_FP;
-	}
-};
+	int length;
+	const char* path = get_amxstring(amx, params[1], 0, length);
+	const char* realpath = build_pathname("%s", path);
 
-enum FileTimeType
-{
-	FileTime_LastAccess = 0,	/* Last access (not available on FAT) */
-	FileTime_Created = 1,		/* Creation (not available on FAT) */
-	FileTime_LastChange = 2,	/* Last modification */
-};
+	AutoPtr<CDirectory> dir(g_LibSys.OpenDirectory(realpath));
 
-static cell AMX_NATIVE_CALL read_dir(AMX *amx, cell *params)
-{
-#ifdef __GNUC__
-	int a;
-	struct dirent *ep;
-	DIR *dp;
-	char* dirname = build_pathname("%s", get_amxstring(amx, params[1], 0, a));
-	a = params[2];
-	
-	if ((dp = opendir (dirname)) == NULL)
-		return 0;
-	
-	seekdir(dp, a);
-	
-	if ((ep = readdir (dp)) != NULL)
+	if (!dir)
 	{
-		cell *length = get_amxaddr(amx, params[5]);
-		*length = set_amxstring(amx, params[3], ep->d_name, params[4]);
-		a = telldir(dp);
-	} else
-		a = 0;
-	
-	closedir (dp);
-	
-	return a;
+		return 0;
+	}
 
-#else
-	int tmp;
-	char *dirname = build_pathname("%s/*", get_amxstring(amx, params[1], 0, tmp));
-	tmp = params[2];
+	cell offset = Max(0, params[2]);
 
-	_finddata_t fd;
-	intptr_t handle = _findfirst(dirname, &fd);
-	
-	if (handle < 0)
-		return 0;
+	dir->NextEntry(&offset);
 
-	++tmp;
-	
-	for (int i = 0; i < tmp; ++i)
+	if (!dir->IsValid())
 	{
-		if (_findnext(handle, &fd) < 0)
-		{
-			tmp = 0;
-			break;
-		}
+		return 0;
 	}
-	
-	// current data in fd
-	cell *length = get_amxaddr(amx, params[5]);						// pointer to the outLen parameter
-	*length = set_amxstring(amx, params[3], fd.name, params[4]);	// set output and outLen parameters
-	_findclose(handle);
 
-	return tmp;
-#endif // __GNUC__
+	const char* entry = dir->GetEntryName();
+
+	size_t entrylen = set_amxstring_utf8(amx, params[3], entry, strlen(entry), params[4] + 1);
+	*get_amxaddr(amx, params[5]) = entrylen;
```

Comment:
Not really fond of combining `get_amxaddr` with the write access like that. Several other instances of this in the file.